compiler/resolution/functionResolution.cpp

```diff
@@ -8278,6 +8278,75 @@ static bool isVecIterator(FnSymbol* fn) {
   return false;
 }
 
+//
+// If the returnSymbol of 'fn' is assigned to from an _array record, insert
+// an autoCopy for that _array. If the autoCopy has not yet been resolved, this
+// function will call 'resolveAutoCopyEtc'.
+//
+// This supports the 'copy-out' rule for returning arrays.
+//
+static void insertUnrefForArrayReturn(FnSymbol* fn) {
+  Symbol* ret = fn->getReturnSymbol();
+  Type* retType = ret->type;
+
+  if (retType == dtUnknown) {
+    for_SymbolSymExprs(se, ret) {
+      if (CallExpr* call = toCallExpr(se->parentExpr)) {
+        if (call->isPrimitive(PRIM_MOVE) && se == call->get(1)) {
+          Type* rhsType = call->get(2)->typeInfo();
+          // BHARSH: Should this also check if fn->retTag != RET_TYPE?
+          //
+          // TODO: Should we check if the RHS is a symbol with 'no auto
+          // destroy' on it? If it is, then we'd be copying the RHS and it
+          // would never be destroyed...
+          if (rhsType->symbol->hasFlag(FLAG_ARRAY) &&
+              isTypeExpr(call->get(2)) == false &&
+              !fn->hasFlag(FLAG_CONSTRUCTOR) &&
+              !fn->hasFlag(FLAG_NO_COPY_RETURN) &&
+              !fn->hasFlag(FLAG_UNALIAS_FN) &&
+              !fn->hasFlag(FLAG_RUNTIME_TYPE_INIT_FN) &&
+              !fn->hasFlag(FLAG_INIT_COPY_FN) &&
+              !fn->hasFlag(FLAG_AUTO_COPY_FN) &&
+              !fn->hasFlag(FLAG_IF_EXPR_FN) &&
+              !fn->hasFlag(FLAG_RETURNS_ALIASING_ARRAY)) {
+            SymExpr* copiedRHS = NULL;
+            if (isSymExpr(call->get(2))) {
+              copiedRHS = toSymExpr(call->get(2)->remove());
+            } else if (CallExpr* rhsCall = toCallExpr(call->get(2))) {
+              if (rhsCall->isPrimitive(PRIM_DEREF)) {
+                // BHARSH 2017-02-15:
+                // ReturnByRef::updateAssignmentsFromRefTypeToValue in the
+                // callDestructors pass would otherwise attempt to insert an
+                // additional autoCopy when it sees the PRIM_DEREF. I observed
+                // that the copy would not be destroyed.
+                copiedRHS = toSymExpr(rhsCall->get(1)->copy());
+                call->get(2)->remove();
```

Comment:
I'm pretty sure the problem still exists, but I'm not confident enough in my knowledge of functionResolution and callDestructors to say that this is a good solution.